### PR TITLE
Add isalpha, istitle, and title APIs in str

### PR DESCRIPTION
**Pull request by Agent-Hellboy**

### CODE REVIEW

**Comment by certik** [comment]
@Agent-Hellboy thanks, looks good. Make sure to update reference tests using `./run_tests -u`.

File: integration_tests/test_str_01.py [hunk 1]
```diff
@@ -56,5 +80,5 @@ def check():
     test_str_slice()
     test_str_repeat()
     test_constant_str_subscript()
-
+    test_str_title()
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/test_str_01.py:83]: Let's test other two functions as well, `test_str_istitle` and `test_str_isalpha`

File: src/runtime/lpython_builtin.py [hunk 2]
```diff
@@ -764,6 +764,70 @@ def _lpython_str_upper(x: str) -> str:
...
 
+@overload
+def _lpython_str_isalpha(s: str) -> bool:
+    char: str
+    for char in s:
+        if 'a' <= char and char <= 'z':
+            continue
+        if 'A' <= char and char <= 'Z':
+            continue
+        return False
```
Comment (Thirumalai-Shaktivel) [inline src/runtime/lpython_builtin.py]:
```suggestion
        if 'a' <= char and char <= 'z':
            continue
        elif 'A' <= char and char <= 'Z':
            continue
        else:
            return False
```

File: src/runtime/lpython_builtin.py [hunk 3]
```diff
@@ -764,6 +764,72 @@ def _lpython_str_upper(x: str) -> str:
     return res
 
 
+@overload
+def _lpython_str_isalpha(s: str) -> bool:
+    char: str
```
Comment (Thirumalai-Shaktivel) [inline src/runtime/lpython_builtin.py:769]: Here is the cause for the CI failure: https://github.com/lcompilers/lpython/issues/2359

I think renaming it or fixing the bug would resolve the CI failure.

**Comment by Thirumalai-Shaktivel** [comment]
Otherwise, this PR looks great!

**Comment by Thirumalai-Shaktivel** [comment]
@Agent-Hellboy you can try `cd integration_tests && ./run_tests.py -b c` to test `c` backend locally.

**Comment by Agent-Hellboy** [comment]
> @Agent-Hellboy you can try `cd integration_tests && ./run_tests.py -b c` to test `c` backend locally.

okay , sure

**Comment by Thirumalai-Shaktivel** [comment]
I'm not sure what's causing the issue. Try debugging it logically and create an issue for it.
You can use `lpython a.py --show-c` to view the c code and `lpython a.py --backend=c` to compile LPython using the C backend.

**Comment by Agent-Hellboy** [comment]
> I'm not sure what's causing the CI failure. Try debugging it logically and create an issue for it. You can use `lpython a.py --show-c` to view the c code and `lpython a.py --backend=c` to compile LPython using the C backend.

sure, I will try to do it today

**Comment by Agent-Hellboy** [comment]
@Thirumalai-Shaktivel, how can I link libasr library I generated main.c  which has included some header file form libasr, but i don't know how to link libasr to compile it with gcc.

**Comment by Thirumalai-Shaktivel** [comment]
Can you please share some details of where you are trying to link libasr?

**Comment by Thirumalai-Shaktivel** [comment]
If you trying to compile C code generated using `--show-c`, try specifying `--backend=c` to lpython. It links all the libraries.
Like, `lpython a.py --backend=c` will compile the generated c code using gcc or clang.

**Comment by Agent-Hellboy** [comment]
> If you are trying to compile C code generated using `--show-c`, try specifying `--backend=c` in lpython. It links all the libraries.

okay

**Comment by certik** [comment]
Once it is ready for review, go ahead and click on the "Ready for review" button.

File: src/runtime/lpython_builtin.py [hunk 4]
```diff
@@ -685,6 +685,71 @@ def _lpython_str_join(s:str, lis:list[str]) -> str:
         res += s + lis[i]
     return res
 
+@overload
```
Comment (certik) [inline src/runtime/lpython_builtin.py:688]: Why do these need `@overload`? I think they are not overloaded, are they?

**Review by certik — APPROVED** [review]
I think this is fine. I left one minor comment.

@Thirumalai-Shaktivel would you mind please reviewing this as well, to make sure I didn't miss anything? If it looks good, you can merge it.

File: integration_tests/test_str_01.py [hunk 5]
```diff
@@ -88,5 +112,7 @@ def check():
     test_str_join_empty_str()
     test_str_join_empty_list()
     test_constant_str_subscript()
-
+    test_str_title()
+    test_str_istitle()
+    test_str_isalpha()
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/test_str_01.py:117]:
```suggestion
    test_str_isalpha()

```

File: src/runtime/lpython_builtin.py [hunk 6]
```diff
@@ -685,6 +685,71 @@ def _lpython_str_join(s:str, lis:list[str]) -> str:
...
+    for ch in s:
+        ch_ord: i32 = ord(ch)
+        if 65 <= ch_ord and ch_ord <= 90:
+            continue
+        if 97 <= ch_ord and ch_ord <= 122:
+            continue
+        return False
+    return True
+
+@overload
```
Comment (Thirumalai-Shaktivel) [inline src/runtime/lpython_builtin.py:700]:
```suggestion
```

File: src/runtime/lpython_builtin.py [hunk 7]
```diff
@@ -685,6 +685,71 @@ def _lpython_str_join(s:str, lis:list[str]) -> str:
...
+                capitalize_next = False
+            else:
+                result += chr(ord(ch) + ord('a') - ord('A'))
+        else:
+            result += ch
+            capitalize_next = True
+
+    return result
+
+@overload
```
Comment (Thirumalai-Shaktivel) [inline src/runtime/lpython_builtin.py:725]:
```suggestion
```

File: integration_tests/test_str_01.py [hunk 8]
```diff
@@ -37,6 +37,30 @@ def test_str_slice():
     # TODO:
     # assert a[0:5:-1] == ""
 
+def test_str_isalpha():
+    a:str = "helloworld"
+    b:str = "hj kl"
+    res: bool = a.isalpha()
+    res2: bool = b.isalpha()
+    assert res == True 
+    assert res2 == False
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/test_str_01.py:46]: Let's add a test for empty string.
Also, add "a12(){}A".

File: integration_tests/test_str_01.py [hunk 9]
```diff
@@ -37,6 +37,30 @@ def test_str_slice():
...
+   
+def test_str_title():
+    a:str = "hello world"
+    b:str = "hj'kl"
+    res: str = a.title()
+    res2: str = b.title()
+    assert res == "Hello World" 
+    assert res2 == "Hj'Kl"
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/test_str_01.py:54]: Same test empty string.
Also, add "hELlo wOrlD", "{Hel1o}world" or something similar

File: integration_tests/test_str_01.py [hunk 10]
```diff
@@ -37,6 +37,30 @@ def test_str_slice():
...
+
+def test_str_istitle():
+    a:str = "Hello World"
+    b:str = "Hj'kl"
+    res: bool = a.istitle()
+    res2: bool = b.istitle()
+    assert res == True
+    assert res2 == False
```
Comment (Thirumalai-Shaktivel) [inline integration_tests/test_str_01.py:62]: Add test for empty string.
Also add " Hello", "HeLlo" or something similar

**Comment by Thirumalai-Shaktivel** [comment]
Rest looks great, I left minor improvements.
Also, we have to resolve the conflicts.

File: src/runtime/lpython_builtin.py (hunk 6 above)
Comment (Thirumalai-Shaktivel) [reply src/runtime/lpython_builtin.py:700]: Let's overload from all three functions

File: src/runtime/lpython_builtin.py [hunk 11]
```diff
@@ -685,6 +685,91 @@ def _lpython_str_join(s:str, lis:list[str]) -> str:
...
+    return True
+
+
+<<<<<<< HEAD
+<<<<<<< HEAD
+=======
+
+>>>>>>> 5c5bcc620 (Fix merge conflict)
+=======
+>>>>>>> f7fcb9835 (Fix test references)
```
Comment (Thirumalai-Shaktivel) [inline src/runtime/lpython_builtin.py:772]: Remove merge conflict tags
```suggestion
        elif ch.isalpha() and (ord('A') <= ord(ch) and ord(ch) <= ord('Z')):
            if word_start:
                word_start = False
            else:
                return False  # Found an uppercase character in the middle of a word
        elif ch.isalpha() and (ord('a') <= ord(ch) and ord(ch) <= ord('z')):
            if word_start:
                return False  # Found a lowercase character in the middle of a word
            word_start = False
        else:
            word_start = True

    return True

```

File: integration_tests/test_str_01.py (hunk 8 above)
Comment (Thirumalai-Shaktivel) [reply integration_tests/test_str_01.py:48]: I think you forgot to add `d: str = ""` and test `d.isalpha()`, `d.istitle()`.

File: src/runtime/lpython_builtin.py (hunk 11 above)
Comment (Agent-Hellboy) [reply src/runtime/lpython_builtin.py]: yes

File: integration_tests/test_str_01.py (hunk 10 above)
Comment (Agent-Hellboy) [reply integration_tests/test_str_01.py:76]: Thanks, this was an edge case 
CPython `istitle and title` is not clear in documentation 
updated the code according to the response from cpython
```
>>> l="\tHello"
>>> l.istitle()
True
>>> l="\thello"
>>> l.istitle()
False
```

File: integration_tests/test_str_01.py (hunk 8 above)
Comment (Agent-Hellboy) [reply integration_tests/test_str_01.py:50]: Added, Again I needed to make changes in `istitle` code. I am feeling confident with `istitle` and `title`, I will revisit them while porting it in CPP as we have to implement these in CPP as others `is*` are implemented in both cpp and Python.

**Comment by Agent-Hellboy** [comment]
merge it for now, from now on, I will pick test cases from CPython
test everything from https://github.com/Agent-Hellboy/cpython/blob/main/Lib/test/test_str.py#L696 
https://github.com/Agent-Hellboy/cpython/blob/main/Lib/test/test_str.py#L734

File: integration_tests/test_str_01.py (hunk 5 above)
Comment (certik) [reply integration_tests/test_str_01.py:139]: @Agent-Hellboy please address this one.

**Review by certik — APPROVED** [review]
